### PR TITLE
job-manager: cache jobspec for use by priority plugin/scheduler

### PR DESCRIPTION
**Pull request by garlick**
This PR modifies the job manager so that it caches the jobspec for each job, if provided by the ingest module.

The ingest module is modified to pass jobspec along with the job submission request, after redacting the environment to make it more compact.

A `job-manager.getattr` RPC is added for general testing support, and a sharness ensures that jobspec is indeed cached, and without environment.

This prepares for the possibility of a priority plugin requiring the resource section + duration to calculate the "size" of a request, as discussed in #3336

### CODE REVIEW

**Comment by grondo** [comment]
Restarted the asan test which hung here:
```
Error: Detected 1 missing missing tests:

t/t2233-job-info-security.t
t/t2233-job-info-security.log
  2020-12-02T19:48:09.910729Z broker.err[0]: rc1.0: Setting fake resource.R={"version": 1, "execution": {"R_lite": [{"rank": "0-3", "children": {"core": "0-1"}}], "starttime": 0.0, "expiration": 0.0, "nodelist": ["fv-az139-[562,562,562,562]"]}}
t/t2233-job-info-security.output
  expecting success: 
          flux jobspec --format json srun -N1 sleep 300 > sleeplong.json
  
  ok 1 - job-info: generate jobspec for simple test job
  job-info: generate jobspec for simple test job
  
  expecting success: 
          jobid=$(submit_job) &&
          flux job eventlog $jobid
  
```

File: src/modules/job-ingest/job-ingest.c [hunk 1]
```diff
@@ -103,6 +103,7 @@ struct job {
 
     char *jobspec;      // jobspec, not \0 terminated (unwrapped from signed)
     int jobspecsz;      // jobspec string length
+    json_t *jobspec_obj;
```
Comment (chu11) [inline src/modules/job-ingest/job-ingest.c:106]: i'd add a comment noting that this is not guaranteed to be a decoded representation of `job->jobspec`, b/c you've redacted environment stuff.  Could trip up someone going through code in the future.

Comment (garlick) [reply src/modules/job-ingest/job-ingest.c:106]: Thanks, agreed!  I was having the same thought.

**Comment by garlick** [comment]
Just pushed some more changes that modify the `sched.alloc` request to include jobspec, now that the job manager has it.  Although the environment is redacted, I assume that should not be required by the scheduler.

This change is handled inside libschedutil, so no change is needed to the scheduler code.

I used @grondo's `bulksubmit.py` test to check performance, but was unable to detect any improvement.  It _should_ reduce the KVS load somewhat in a high throughput scenario. which can't be a bad thing.

flux-framework/rfc#279 was submitted to capture the protocol change in RFC 27

**Comment by garlick** [comment]
(Force pushed a bit of commit tidying - including adding a reference to the bug that this closes, and reducing the size of the diff slightly in a couple of commits)

File: src/modules/job-ingest/job-ingest.c (hunk 1 above)
Comment (garlick) [reply src/modules/job-ingest/job-ingest.c:106]: I called the variable `job->jobspec_redacted` in the job manager, thinking that was where maximum confusion might occur, then just noticed your comment was on the code in job-ingest.  Maybe I should add a comment there.

File: t/t2212-job-manager-jobspec.t [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+test_expect_success 'stop the queue for this test' '
+	flux queue stop
+'
+
+test_expect_success 'create simple jobspec' '
+	flux mini submit --dry-run hostname >simple.json
+'
+
+test_expect_success HAVE_JQ 'jobspec contains environment' '
+	jq -e -j .attributes.system.environment <simple.json >env.json
```
Comment (chu11) [inline t/t2212-job-manager-jobspec.t:26]: nit: i don't think you need `-j` for these tests, since you're not comparing output

**Review by chu11 — COMMENTED** [review]
LGTM, just noticed 1 nit.

File: t/t2212-job-manager-jobspec.t (hunk 2 above)
Comment (garlick) [reply t/t2212-job-manager-jobspec.t]: OK, fixed that and forced a push.

**Review by chu11 — APPROVED** [review]
LGTM

**Comment by garlick** [comment]
Thanks!  I'll go ahead and set mwp.

**Pull request merged**